source/FreeRTOS_IP_Utils.c

```diff
@@ -861,6 +861,22 @@ void prvProcessNetworkDownEvent( struct xNetworkInterface * pxInterface )
          * treat network down as a "delivery problem" and flush the ARP cache for this
          *  interface. */
         FreeRTOS_ClearARP( pxEndPoint );
+
+        #if ( ( ipconfigUSE_DHCP != 0 ) || ( ipconfigUSE_DHCPv6 != 0 ) )
+            if( END_POINT_USES_DHCP( pxEndPoint ) )
+            {
+                /* Stop the DHCP process for this end-point. */
+                vIPSetDHCP_RATimerEnableState( pxEndPoint, pdFALSE );
```

Comment:
When you try to end a DHCP procedure just by disabling the timer, there is a risk that the DHCP socket won't be closed.
In DHCP/IPv4, the socket has a reference count called `xDHCPSocketUserCount`. It should be closed by calling `prvCloseDHCPSocket`.
For DHCPv6, this is much the same, the socket is called `xDHCPv6Socket`, and it is protected by `xDHCPv6SocketUserCount`, and closed by calling `prvCloseDHCPv6Socket()`.

Reply:
I added `vDHCPStop` and `vDHCPv6Stop` functions that, besides disabling the timer, also call `prvCloseDHCPSocket` to attempt to close the socket. 

I also noticed that IPv4 implementation of DHCP is not aware which endpoints opened the socket and calling `prvCloseDHCPSocket` will decrease `xDHCPSocketUserCount` even when the endpoint did not open it, whereas the DHCPv6 implementation stores the socket locally for each endpoint. That's why I decided to change DHCPv4 implementation to be more in line with DHCPv6.